h/views/admin/documents.py

```diff
@@ -0,0 +1,53 @@
+import json
+
+from pyramid.view import view_config, view_defaults
+
+from h.schemas import ValidationError
+from h.schemas.annotation import URLMigrationSchema
+from h.security import Permission
+from h.tasks.url_migration import move_annotations_by_url
+
+
+@view_defaults(route_name="admin.documents", permission=Permission.AdminPage.HIGH_RISK)
+class DocumentsAdminViews:
+    def __init__(self, request):
+        self.request = request
+
+    @view_config(
+        request_method="GET", renderer="h:templates/admin/documents.html.jinja2"
+    )
+    def get(self):
+        return {}
+
+    @view_config(
+        request_method="POST",
+        request_param="update_annotation_urls",
+        require_csrf=True,
+        renderer="h:templates/admin/documents.html.jinja2",
+    )
+    def update_annotation_urls(self):
+        try:
+            url_mappings_json = json.loads(self.request.params["url_mappings"])
+        except ValueError as err:
+            return self.report_error(f"Failed to parse URL mappings: {err}")
+
+        schema = URLMigrationSchema()
+
+        try:
+            url_mappings = schema.validate(url_mappings_json)
+        except ValidationError as err:
+            return self.report_error(f"Failed to validate URL mappings: {err}")
```

Comment:
I'm guessing this is endemic throughout the admin pages, but the way this should really work is something like this:

```suggestion
    @view_config(
        request_method="POST",
        request_param="update_annotation_urls",
        require_csrf=True,
        renderer="h:templates/admin/documents.html.jinja2",
        schema=URLMigrationSchema,
    )
    def update_annotation_urls(self):
        # If we get here then the request body has already been validated.
        validated_json = self.request.validated_json
        
        ...
```

That is: the view function itself doesn't have to pull the `"url_mappings"` param out of the request, deal with the case of a missing param, call `URLMigrationSchema.validate(), or deal with the case of a validation error. That would all be done by a Pyramid view predicate (the `schema=URLMigrationSchema` argument to `@view_config()`) which would send the appropriate error response if necessary or call the view if the request is valid.

We'd probably have to implement a custom view predicate to make this work. But you can see how much shorter it'd make the code of all the admin pages views (and their unit tests) while also making error handling behaviour consistent.